packages/contracts-bedrock/src/libraries/Transient.sol

```diff
@@ -0,0 +1,121 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+/// @title Transient
+/// @notice Library for transient storage.
+library Transient {
+    /// @notice Slot for call depth.
+    ///         Equal to bytes32(uint256(keccak256("transient.calldepth")) - 1).
+    uint256 internal constant CALL_DEPTH_SLOT = 0x7a74fd168763fd280eaec3bcd2fd62d0e795027adc8183a693c497a7c2b10b5c;
+
+    /// @notice Gets the current call depth.
+    /// @return _callDepth Current call depth.
+    function getCallDepth() internal view returns (uint256 _callDepth) {
+        assembly {
+            _callDepth := tload(CALL_DEPTH_SLOT)
+        }
+    }
+
+    /// @notice Increments call depth.
+    function incrementCallDepth() internal {
+        // Get the current call depth
+        uint256 currentCallDepth = getCallDepth();
+
+        // Increment the call depth
+        assembly {
+            tstore(CALL_DEPTH_SLOT, add(currentCallDepth, 1))
+        }
+    }
+
+    /// @notice Decrements call depth.
+    function decrementCallDepth() internal {
+        // Get the current call depth
+        uint256 currentCallDepth = getCallDepth();
+
+        // Decrement the call depth
+        assembly {
+            tstore(CALL_DEPTH_SLOT, sub(currentCallDepth, 1))
+        }
+    }
+
+    /// @notice Gets corresponding transient slot for an arbitrary slot at a given call depth.
+    /// @param _callDepth Call depth to get transient slot for.
+    /// @param _slot      Slot to get transient slot for.
+    /// @return Corresponding transient slot at the given call depth.
+    function getTransientSlot(uint256 _callDepth, bytes32 _slot) internal pure returns (bytes32) {
+        return keccak256(abi.encodePacked(_callDepth, _slot));
+    }
+
+    /// @notice Gets corresponding transient slot for an arbitrary slot at current call depth.
+    /// @param _slot Slot to get transient slot for.
+    /// @return Corresponding transient slot at the current call depth.
+    function getTransientSlotCurrentDepth(bytes32 _slot) internal view returns (bytes32) {
+        return keccak256(abi.encodePacked(getCallDepth(), _slot));
+    }
+
+    /// @notice Sets a value in transient storage at slot at a given call depth.
+    /// @param _callDepth Call depth to get transient slot for.
+    /// @param _slot    Slot to set.
+    /// @param _value   Value to set.
+    function setTransientValue(uint256 _callDepth, bytes32 _slot, uint256 _value) internal {
```

Comment:
There are a lot of functions in the api, making this library complex to consume.

Really it should be:

- `get(bytes32)(bytes32)`
- `set(bytes32, bytes32)`
- `calldepth()(uint256)`
- `increment()`
- `decrement()`

Do we need anything else?

Reply:
internal to `get` + `set` it handles mixing to create the transient key based on the call depth. we don't need to expose that logic externally, otherwise it creates a leaky abstraction. the caller doesn't care where the key is stored, they just care that its stored. does this make sense?

Reply:
good idea!